### PR TITLE
Feature/gradientElements

### PR DESCRIPTION
**Pull request by cyriacbijun**
Takes a prop `gradientList` that takes list of colors as elements of list. Another prop `isEmptyColorGradient` (default:`false`) to set background color as gradient

### CODE REVIEW

**Comment by apgapg** [comment]
Hi @cyriacbijun Gradient will be really a cool add on.
Can you please post a screenshot for seeing visual appearance?

**Comment by cyriacbijun** [comment]
Sure, @apgapg . These are the gradients I took :
```
 const greenGradient = [
      Color.fromRGBO(223, 250, 92, 1),
      Color.fromRGBO(129, 250, 112, 1)
    ];
    const turqoiseGradient = [
      Color.fromRGBO(129, 182, 205, 1),
      Color.fromRGBO(91, 253, 199, 1),
    ];

    const redGradient = [
      Color.fromRGBO(175, 63, 62, 1.0),
      Color.fromRGBO(254, 154, 92, 1),
    ];
```
Case 1 : without gradient background
```
PieChart(
                dataMap: dataMap,
                gradientList: [greenGradient, turqoiseGradient],
                ringStrokeWidth: 10,
                chartRadius: 140,
                chartType: ChartType.ring,
                initialAngleInDegree: 70,
                chartValuesOptions: const ChartValuesOptions(
                  showChartValueBackground: false,
                  showChartValues: false,
                ),
                legendOptions: const LegendOptions(
                  showLegends: false,
                ),
              ),
```
normal_background

Case 2 : All elements gradient
```
PieChart(
                dataMap: dataMap,
                gradientList: [greenGradient, turqoiseGradient,redGradient],
                isEmptyColorGradient: true,
                ringStrokeWidth: 10,
                chartRadius: 140,
                chartType: ChartType.ring,
                initialAngleInDegree: 70,
                chartValuesOptions: const ChartValuesOptions(
                  showChartValueBackground: false,
                  showChartValues: false,
                ),
                legendOptions: const LegendOptions(
                  showLegends: false,
                ),
              ),
```
gradient_background

**Comment by apgapg** [comment]
@cyriacbijun I don't think we need to keep a new param `isEmptyColorGradient`. We can check if gradientColorList is passed by checking not null or empty and automatically superseding it over the color list.
What's your thought?

**Comment by cyriacbijun** [comment]
> 
> 
> @cyriacbijun I don't think we need to keep a new param `isEmptyColorGradient`. We can check if gradientColorList is passed by checking not null or empty and automatically superseding it over the color list. What's your thought?

Well, I had intended to use the param  `isEmptyColorGradient` to enable user to have a single element without any gradient( usually a style element with no particular value in relation to the chart, as seen in my example). This element would use the `emptyColor` param as its Color. In other cases, where all the elements are gradients, this would be `true`. If this is not required let's go ahead without it.
Also, I hadn't used the gradients in the chart Legend element, do we use the 1st color for it or should it be a linear gradient filled box?

**Comment by apgapg** [comment]
@cyriacbijun I agree with your saying.
For Legend, I don't think we really require gradient. The first color would work fine if the colors inside the gradient are similar.
I think we are ready to merge this.

File: lib/src/chart_painter.dart [hunk 1]
```diff
@@ -46,15 +48,19 @@ class PieChartPainter extends CustomPainter {
...
+    this.gradientList,
+    this.isBackgroundColorGradient = false,
   }) {
     _total = values.fold(0, (v1, v2) => v1 + v2);
-    for (int i = 0; i < values.length; i++) {
-      final paint = Paint()..color = getColor(colorList, i);
-      if (chartType == ChartType.ring) {
-        paint.style = PaintingStyle.stroke;
-        paint.strokeWidth = strokeWidth!;
+    if (gradientList == null) {
```
Comment (apgapg) [inline lib/src/chart_painter.dart:55]: Here IMO we should also put empty list check

File: lib/src/chart_painter.dart [hunk 2]
```diff
@@ -79,14 +85,59 @@ class PieChartPainter extends CustomPainter {
       );
     } else {
       _prevAngle = this.initialAngle! * math.pi / 180;
+      final isGradientPresent = gradientList != null;
```
Comment (apgapg) [inline lib/src/chart_painter.dart:88]: Similarly here an empty list check will make it more safe to use

**Review by apgapg — CHANGES_REQUESTED** [review]

**Comment by apgapg** [comment]
@cyriacbijun When `isEmptyColorGradient` is false, the chart will use default `emptyColor` color.
What happens when `isEmptyColorGradient` is true? What would be the `emptyColor` in that case?

Also `isEmptyColorGradient` doesn't look self explanatory, can we find a new name?

**Comment by cyriacbijun** [comment]
> @cyriacbijun When `isEmptyColorGradient` is false, the chart will use default `emptyColor` color. What happens when `isEmptyColorGradient` is true? What would be the `emptyColor` in that case?
> 
> Also `isEmptyColorGradient` doesn't look self explanatory, can we find a new name?

when `isEmptyColorGradient` is true, the first element from the `gradientList` is taken. Another way to avoid this variable is to compare the length of `dataMap` and `gradientList`. If length of `gradientList` is one less, the simple color element automatically takes the default `emptyColor` color. Is this a better way? 
If not, please suggest another name for `isEmptyColorGradient` param. I'll do the requested changes ASAP

**Comment by apgapg** [comment]
@cyriacbijun 
`If length of gradientList is one less, the simple color element automatically takes the default emptyColor color. Is this a better way?`
Yes, this looks better way. The fallback will be empty color.

You can use `useEmptyColorAsGradient` as naming. This looks better in my view. What you say?

**Comment by cyriacbijun** [comment]
> @cyriacbijun `If length of gradientList is one less, the simple color element automatically takes the default emptyColor color. Is this a better way?` Yes, this looks better way. The fallback will be empty color.
> 
> You can use `useEmptyColorAsGradient` as naming. This looks better in my view. What you say?

All right then, We'll completely avoid introducing a new param and use the comparison method. I have updated the PR, please have a look.

**Comment by apgapg** [comment]
@cyriacbijun I think a better way would be to introduce a new `emptyColorGradient`. In this way we need not depend on gradient list to extract empty color. This looks more sensible.

`emptyColor: instance of Color class`
`emptyColorGradient: instance of Gradient class`

**Comment by cyriacbijun** [comment]
@apgapg  I have added the new param, please have a check. `emptyColorGradient` is of type `List<Color>` as that was more flexible, I hope that's fine

**Review by apgapg — APPROVED** [review]
LGTM.

Thank you for such cool addition. @cyriacbijun

**Comment by apgapg** [comment]
v5.1.0 released.

@cyriacbijun Thank you for your contribution

**Pull request merged**